ql/ql/src/queries/style/OmittableExists.ql

```diff
@@ -10,12 +10,22 @@
 
 import ql
 
+class AggregateOrForQuantifier extends AstNode {
+  AggregateOrForQuantifier() {
```

Comment:
Checkout `getConjunctionParent` in `RedundantAssignment.ql`, I think you are trying to solve the same problem I was.   

I wanted to ensure that there were no disjunctions or negative edges between two AST-nodes, and I think you might be trying to do the same thing.  

You could extract the relevant code from `RedundantAssignment.ql` into some `.qll` file.

Reply:
I created `AggregateOrForQuantifier` because aggregations and `for[all|ex]` behave differently when you use an unbound exists variable compared to a don't-care expression.

What you're referring to seems more related to the FN case in the tests, isn't it?

```ql
exists(int i | aPredicate(i) or anotherPredicate(i)) // BAD [NOT DETECTED]
```

Or am I missing something?

Reply:
> I created AggregateOrForQuantifier because aggregations and `for[all|ex]` behave differently when you use an unbound exists variable compared to a don't-care expression.

That is only when the variable is declared by the `for[all|ex]` statement, right?  
Because in that case I think you should change `not use.getParent*() instanceof AggregateOrForQuantifier` to `not existsArgument.getParent() AggregateOrForQuantifier`. 

(And in that case my former comment is not relevant).

Reply:
> That is only when the variable is declared by the for[all|ex] statement, right?

Not really. I've seen changes in behavior introduced by diffs like the following:

```diff
predicate loopExitGuard(LoopStmt loop, Expr cond) {
-  exists(ConditionBlock cb, boolean branch |
+  exists(ConditionBlock cb |
    cond = cb.getCondition() and
    cond.getEnclosingStmt().getEnclosingStmt*() = loop.getBody() and
-    forex(Stmt exit | loopExit(loop, exit) | cb.controls(exit.getBasicBlock(), branch))
+    forex(Stmt exit | loopExit(loop, exit) | cb.controls(exit.getBasicBlock(), _))
  )
}

Reply:
See my updated answer. I was wrong in my first comment.

Reply:
Why are you repeating the `One e | not always(e)` case?

Shouldn't it be like the following (in which case 2 results appear, `not never(_)` and `One e | not never(e)`)?

```ql
predicate test(string msg) {
  exists(One e | not always(e)) and
  msg = "One e | not always(e)"
  or
  exists(One e | not never(e)) and
  msg = "One e | not never(e)"
  or
  not always(_) and
  msg = "not always(_)"
  or
  not never(_) and
  msg = "not never(_)"
}
```

> Another thing you don't handle by just using getParent*() is if the for[all|ex] is "above" the exists.

That's true. I'll try your approach.

Reply:
> Why are you repeating the One e | not always(e) case?

My mistake, yeah my example was bad, sorry.  
But I'm quite sure that `not` (and other negative edges) can cause similar issues in other (less trivial) cases.

Reply:
Ok, back from holidays and trying to get this done. I generalized a predicate like the following and put in a shared QLL so that it can be used by this query and `RedundantAssignment.ql`:

```ql
private AstNode getConjunctionParentRec(AstNode p) {
  result = p
  or
  result = getConjunctionParentRec(p).getParent() and
  not result instanceof Disjunction and
  not result instanceof IfFormula and
  not result instanceof Implication and
  not result instanceof Negation and
  not result instanceof Predicate and
  not result instanceof FullAggregate and
  not result instanceof Forex and
  not result instanceof Forall
}
```

My concern is that `p` is no longer bound, and this predicate will become absurdly large. If I'm not mistaken and this really is a performance problem, the only solution I can think of is creating a parameterized module that receives the predicate that binds `p` as a parameter, but it sounds a little overkill...

WDYT?

Reply:
Also, `RedundantAssignment.ql` doesn't seem to have any tests, so I don't feel very comfortable modifying it.

Reply:
Ok, thanks for the guidance! Done in ae8c75a.